gpytorch/priors/lkj_prior.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import math
+import torch
+from torch.distributions.constraints import positive_definite
+from gpytorch.priors.prior import Prior
+
+
+class LKJPrior(Prior):
+    """LKJ prior over n x n (positive definite) correlation matrices
+
+    pdf(Sigma) ~ |Sigma| ^ (eta  - 1)
+
+    where eta > 0 is a shape parameter and n is the dimension of the
+    correlation matrix.
+    """
+
+    def __init__(self, n, eta=1):
+        if eta <= 0:
+            raise ValueError("Shape parameter of LKJ prior must be positive")
+
+        if n < 1:
+            raise ValueError("Dimension n must be a positive integer")
+        super(LKJPrior, self).__init__()
+        self.register_buffer("eta", torch.tensor(eta))
+        self.register_buffer("n", torch.tensor(n))
+        # Normalization constant
+        # Reference: Bayesian Data Analysis, 3rd ed., Gelman et al., p. 576
+        i = torch.arange(n).float()
+        C = (
+            torch.sum((2 * eta - 2 + i) * i) * math.log(2) +
+            n * torch.sum(2 * torch.lgamma(i / 2 + 1) - torch.lgamma(i + 2))
+        )
+        self.register_buffer("C", C)
+        self._log_transform = False
+
+    def _log_prob(self, parameter):
+        if not is_valid_correlation_matrix(parameter):
+            raise ValueError("Input is not a valid correlation matrix")
+        return self.C + (self.eta - 1) * torch.log(torch.det(parameter))
```

Comment:
`torch.log(torch.det(parameter))` is probably fine for very small matrices, but in general is not a good way to compute log determinants for positive definite matrices. `logdet = parameter.potrf().diag().log().sum() * 2` is much better. Example:

```python
mat_root = torch.randn(50, 50)
covar_mat = mat_root.t().matmul(mat_root)
log_det_bad = torch.log(torch.det(covar_mat))
log_det_good = covar_mat.potrf().diag().log().sum() * 2
print(log_det_bad) # Outputs tensor(inf)
print(log_det_good) # Outputs log | covar_mat |
```